### DISCUSSION TITLE
Is there a plugin to crash when dataset methods are ignored?

### DISCUSSION THREAD

**Discussion by fdr** [jeremyevans/sequel#2208] [General]
Hello everyone,

Contrast these two situations:

> DB["select * from pg_class"].where { oid > 7 }.sql
=> "select * from pg_class"

> DB[:pg_class].where { oid > 7 }.sql
=> "SELECT * FROM \"pg_class\" WHERE (\"oid\" > $1::int4); [7]"

I seem to remember figuring out in the past that, if I wanted this to do what was expected, I wanted the "implicit_subquery" plugin. But, I felt the amount of bloating to enable `implicit_subquery` all the time was perhaps excessive.

Now, it occurs to me: something that seems better would be a plugin that crashes/raises an error when dataset method chaining is ignored.

Is there a plugin like that?

**Comment by jeremyevans** [comment]
It would be fairly easy to make a copy of the implicit_subquery plugin that raises instead of using an implicit subquery.  Is that what you want?  It's just as bloating as implicit_subquery, but honestly, I don't think the amount of bloating is excessive (was it a bottleneck in your application)?

For a general case, it's expected that Sequel dataset methods will only use the parts of the dataset they are expected to use, and not check for parts they are not intended to use:

```ruby
  ds = DB[:table].insert_ignore
  ds.insert(:a=>1) # insert_ignore setting used
  ds.all # insert_ignore setting not used
```

**Comment by fdr** [comment]
I was thinking more it makes the queries themselves a bit longer. It's not so much about system overhead, as the legibility of logs and the like:

```
> puts DB["select * from pg_class"].where { oid > 7 }.sql
SELECT * FROM (select * from pg_class) AS "t1" WHERE ("oid" > $1::int4)
``` 

So far, I've done some small restructuring to do more raw SQL stuff to avoid subquery wrapping. I was thinking I would use  `from_self` on my own when necessary. But, on and off in the last few months, it occurred to me that there is some danger of ignored conditions on raw SQL queries that could cause an information disclosure problem. I never got around to writing my concern until now.

Perhaps I should enable `implicit_subquery` on those grounds alone...

**Reply by jeremyevans** [reply]
If you plan to call dataset query methods on a dataset that uses raw SQL, I would definitely recommend using the implicit_subquery plugin.  It's true that it makes the queries longer as it forces a subquery, but that shouldn't cause a problem.

In general, my recommendation would be to only use a raw SQL dataset in cases where you are immediately calling a dataset method that returns results.  It's definitely a bad idea to pass a raw SQL dataset to other methods that are not designed to expect a raw SQL dataset (or to chain directly as in your example), if you are not using the implicit_subquery plugin.  If you want to avoid the implicit_subquery overhead, then manually calling `from_self` works fine too, but you need to use that consistently.

**Reply by fdr** [reply]
Yeah,  that is what we do, but it occurs to me that the margin for error is awfully small, and silent, we bumped into it by accident. I will enable implicit_subquery in the interest of safety.

**Reply by fdr** [reply]
So, I enabled implicit subquery, and found something interesting:

https://github.com/ubicloud/ubicloud/pull/1900

It looks like `DB["sql-string-here"].get` wraps things when it probably shouldn't, whereas `DB["sql-string-here"].first.fetch(...)` works as expected. In the diff, I have a couple of more tailored uses (where I need the value), otherwise I use it as the shortest way to say, run a utility statement as a side effect (e.g. `GRANT`)

Is the difference in behavior expected? `.get` I think is more elegant for single-element returns.

All in all,  I would still prefer to wrap explicitly and get a crash otherwise I think, since composition might get a bit weird from introduced identifiers. At least I'd have a pause to consider if I should rewrite code in the DSL to avoid wraps if I need more composition, but having implicit_subquery is probably still better than the status quo.

**Reply by jeremyevans** [reply]
`get` wraps because it internally calls `select`.  In your example, with no arguments, you get the equivalent of `SELECT * FROM (SELECT ...)`, so there is no difference in the selected columns.  Maybe we could have `implicit_subquery` handle `get` without arguments specially?

It looks like `first` does not wrap because it uses `clone(:limit=>1)` instead of `limit(1)` internally.  We could change that so that it will set a limit.  Unfortunately, then you get into issues with subqueries being unordered in SQL.  I'm guessing such a change would break backwards compatibility, at least on some databases.

If you want to avoid the `get` issue, switch to using `single_value`.  That's more appropriate as you aren't trying to select an expression, just return the already selected value.

**Reply by fdr** [reply]
> get wraps because it internally calls select. In your example, with no arguments, you get the equivalent of SELECT * FROM (SELECT ...), so there is no difference in the selected columns. Maybe we could have implicit_subquery handle get without arguments specially?

Perhaps, although, it's not clear to me which of the methods I should use for utility statements. "get" could have been "all" or "first" or any number of things. So maybe a special path is not necessary, if there is an idiom of how to run statements where the records that come out are not that interesting (do GRANT statements even yield records? My memory is a bit hazy on this one)

I can use `single_value`, I didn't realize `get` was related to non-trivial expressions. It's a bit of a mouthful, suggesting it is intended to be out of the way.

**Reply by jeremyevans** [reply]
Most utility statements such as `GRANT` do not return an important result, so it's better to use `Database#run` instead of creating a dataset and calling `get`/`single_value`:  `DB.run("GRANT ...")`

For datasets that return a result, if you use raw SQL, it's best to call one of the following methods directly on the dataset with raw SQL:

* `single_value!` # one row, one column
* `single_record!` # one row
* `all` # all rows as hashes
* `map` # array of values or array of arrays of values

I should update the documentation (https://sequel.jeremyevans.net/rdoc/files/doc/querying_rdoc.html#label-Custom+SQL) to better discuss this and the general issues with raw SQL datasets.

**Reply by fdr** [reply]
I think the problem with DB.run has to do with parameters.

This, for example, doesn't work:

```ruby
    db.run("GRANT CREATE ON SCHEMA public TO ?", ph_user.to_sym)
```

(and the documentation suggests it should not work, since run only takes sql and an option hash).

In principle I can mangle stuff myself with quote ident, etc, but by the time I'm done hacking the string together, one wonders why not stick with the (somewhat weird) alternate, which stays in the notation of Sequel's quoting rules that we use frequently:

```ruby
    db["GRANT CREATE ON SCHEMA public TO ?", ph_user.to_sym].first
```

**Reply by jeremyevans** [reply]
Good point.  It's by design that `Database#run` does not accept placeholders.  It is possible to combine them:

```ruby
  db.run(db["GRANT CREATE ON SCHEMA public TO ?", ph_user.to_sym].sql)
```

Maybe we should add an extension for `Dataset#run` so you can do?:

```ruby
  db["GRANT CREATE ON SCHEMA public TO ?", ph_user.to_sym].run
```

**Reply by fdr** [reply]
I think adding `.run` makes sense. It's has a meaningful symmetry with `db.run`, it scans a bit more authoritatively that the result set is not checked vs `.first`, or `.all`. You can imagine this also being useful with INSERT/UPDATE/DELETE as well. Though in principle these can have interesting return values (whether the count is interesting, or with `RETURNING`), for many cases, the error/exceptions are all the programmer needs.

**Reply by jeremyevans** [reply]
OK.  I'll add a dataset_run extension for that in 5.85 (too late for 5.84).  Thanks for the discussion!

**Reply by jeremyevans** [reply]
I've pushed some commits based on this discussion:

* Added dataset_run extension (26a2243ef8f3c8958bed29f23399085872e2442e)
* Make Dataset#get without arguments run query directly (without subquery) when using implicit_subquery extension (45b7f370709c476bc4b7e397fceaa8be0fcb6f99)
* Updated documentation for Custom SQL section of querying guide (0f4cce2aed9a2a74788283d6ff02e7df98a39d05)

Thanks for the feedback and discussion!

**Reply by fdr** [reply]
thanks for your constant and skilled maintenance and design of Sequel (and the others)
